rfc/0067-extended-sdk-observability-rto-v2.md

```diff
@@ -0,0 +1,1144 @@
+
+## TracerException
+
+This exception is raised as a wrapper around all `RequestTracer` implementation errors (i.e. if OpenTelemetry tracer fails, it will bubble up wrapped in that `TracerException`).
+
+## MeterException
+
+This exception is raised as a wrapper around all Meter implementation errors (i.e. if OpenTelemetry meter fails, it will bubble up wrapped in that `MeterException`).
+
+# Cluster labels
+
+From 7.6.4, the server publishes `clusterUUID` and `clusterName` fields in the config, which the SDKs will export in the tags/attributes of all metrics and spans.
+
+If this information is not available for any reason (generally this will only be if the cluster has not upgraded fully to 7.6.4), then the SDK should simply not include these tags/attributes.  (With the previously stated exception of Micrometer.)
+
+To aid performance, the SDK is allowed to cache these fields for up to (an arbitrarily chosen) 10 minutes.  This ensures that any cluster name changes (expected to be rare) are eventually picked up.
+
+# Compatibility with older semantic conventions
+
+SDKs that implemented earlier revisions of this RFC will have implemented semantic conventions that were considered 'Experimental' by OpenTelemetry.
+
+In order to support the stable OpenTelemetry semantic conventions, SDKs should report operation latencies to the `db.client.operation.duration` metric, instead of `db.couchbase.operations`. In addition, the following attributes have been renamed, removed, or added by the migration to the stable semantic conventions.
+
+| | Experimental (v0) | Stable (v1) | Note |
+| -- | ------------ | ------ | ---- |
+| System name | db.system | db.system.name | |
+| Bucket name | db.name | db.namespace | |
+| Query statement | db.statement | db.query.text | |
+| Operation name | db.operation | db.opearation.name | |
+| Error | outcome | error.type | (1) |
+| OSI transport layer | net.transport | network.transport | (2) |
+| Local hostname/IP | net.host.name | Removed | |
+| Local port | net.host.port | Removed | |
+| Server hostname/IP | net.peer.name | server.address | |
+| Server port | net.peer.port | server.port | |
+| Peer address | Did not exist | network.peer.address | |
+| Peer port | Did not exist | network.peer.port | |
+
+**Notes:**
+
+1. In v0, one of the possible values was "Success". In v1, the attribute should be omitted if there was no error.
+2. In v0, the value was "TCP.IP". In v1, it should be "tcp" instead.
+
+Existing implementations should not change the conventions that they emit by default. `OpenTelemetryRequestTracer` and `OpenTelemetryRequestMeter` should recognise the `OTEL_SEMCONV_STABILITY_OPT_IN` and emit the new stable conventions on an opt-in basis, as outlined in the [OpenTelemetry Database Semantic Conventions v1.37 document](https://github.com/open-telemetry/semantic-conventions/blob/v1.37.0/docs/database/database-spans.md):
```

Comment:
I'm undecided if this is the right approach. Should we support sending the new semantic conventions to the `Meter`/`Tracer` interfaces or only implement the conversion to the new conventions within the OpenTelemetry-specific concrete implementations?

Having it all in the OpenTelemetry-specific implementations means that the SDK is agnostic to OpenTelemetry conventions. However, it could cause confusion if the attributes we call the interfaces with attributes that have names that are Otel-like, but are not quite following the current Otel spec. Users implementing the interfaces on their own might also want to get the new semantic conventions without having to do extra work on their side.

Reply:
I think I'm leaning towards changing this to say that the new semantic conventions should be implemented within the SDK, that means all implementations of the `RequestTracer`, `Meter` interfaces can benefit & it's more consistent with the fact that it's the SDK itself implementing the conventions, and the `OpenTelemetryRequestTracer` and `OpenTelemetryRequestMeter` only relay the metrics/spans to an Otel-compatible backend.

Reply:
Updated based on the above.